Stabilize 29 RISC-V target features (`riscv_ratified_v2`)

a4lg opened the pull request:
This commit stabilizes RISC-V target features with following constraints:

*   Describes a ratified extension.
*   Implemented on Rust 1.88.0 or before.  
    Waiting for four+ version cycles seems sufficient.
*   Does not disrupt current rustc's target feature (cf. rust-lang/rust#140570) + ABI (cf. rust-lang/rust#132618) handling.  
    It excludes `E` and all floating point-arithmetic extensions.  The `Zfinx` family does not involve floating point registers but not stabilizing for now to avoid possible confusion between the `F` extension family.
*   Not vector-related (floating point and integer).
    While integer vector subsets should not cause any ABI issues (as they don't use ABI-dependent floating point registers), we need to discuss before stabilizing them.
*   Supported by the lowest LLVM version supported by rustc (LLVM 20).

List of target features to be stabilized:

1.  `b`
2.  `za64rs` (no-RT)
3.  `za128rs` (no-RT)
4.  `zaamo`
5.  `zabha`
6.  `zacas`
7.  `zalrsc`
8.  `zama16b` (no-RT)
9.  `zawrs`
10.  `zca`
11. `zcb`
12. `zcmop`
13. `zic64b` (no-RT)
14. `zicbom`
15. `zicbop` (no-RT)
16. `zicboz`
17. `ziccamoa` (no-RT)
18. `ziccif` (no-RT)
19. `zicclsm` (no-RT)
20. `ziccrse` (no-RT)
21. `zicntr`
22. `zicond`
23. `zicsr`
24. `zifencei`
25. `zihintntl`
26. `zihintpause`
27. `zihpm`
28. `zimop`
29. `ztso`

Of which, 20 of them (29 minus 9 "no-RT" target features) support runtime detection through `std::arch::is_riscv_feature_detected!()`.

Corresponding PR for the Reference: rust-lang/reference#1987

<!-- homu-ignore:start -->

-----

# Description in older revision(s)

> *   Not vector-related (including not FP-related ones).  
>     No ABI issues will not occur unless floating point registers are involved.  But for integer-only vector ones, we'll be stabilizing in another time.

The original text is rewritten to avoid confusion as per @RalfJung's request.

> Of which, 19 of them support runtime detection through `std_detect`.

This is rewritten to correlate this PR with a public API (instead of an internal crate).

r? @Amanieu
@rustbot label +O-riscv
<!-- homu-ignore:end -->

a4lg commented:
I expect that this PR lands on the version 1.92 cycle (I chose extensions with relatively low risk factors) but happy if it could be merged in the version 1.91 cycle.

Amanieu commented:
LGTM, these are all officially ratified RISC-V extensions.

@rfcbot merge

a4lg commented:
# Consideration: Feature Naming (for `std_detect`)

The only initial intent of `riscv_ratified_v2` (which feels a bit *blunt*) was to avoid crashing with existing `riscv_ratified`[^featstable] and later noticed that x86 uses various feature names on stabilization.

[^featstable]: The same feature name does not allow multiple stabilized versions.

I considered the similar after the initial submission but didn't for several reasons:

1.  Many extensions to be stabilized are general purpose (even some of them are split from an early version of the `I` base ISA) and clear distinction will confuse the users between plain `riscv_ratified` (it's just a collection of extensions where runtime detection of them is stabilized in 1.78.0 and contains *rich* extensions like scalar cryptography).
2.  Considering future extensions to be stabilized in Rust and the history (some RISC-V extensions are *split* from others and semantic naming may not work in the near future), either of them seems inevitable:
    1.  Use the extension name itself or similar or
    2.  Use blunt version schemes like in this PR.

After consideration, I prefer *blunt* naming scheme (pure versioning regardless of the category) for runtime detection of RISC-V extensions.

# Consideration: Deferring to Version 1.92 Cycle may Have its Benefit

Although I'll rush for the version 1.91 cycle if possible, I have to say that deferring this might be an option.

I'm talking about #145071, which raises the minimum LLVM version to 20.  After this PR is merged, I think we can safely introduce the `Zacas` extension without `riscv_ratified_v3` (I excluded `Zacas` from this PR because of this minimum LLVM version issue).

So, if we don't make it to the version 1.91 cycle, waiting for #145071 might be an option.

a4lg commented:
@traviscross Thanks for pointing out!

I opened the PR rust-lang/reference#1987 to align with this.

RalfJung commented:
> No ABI issues will not occur unless floating point registers are involved. But for integer-only vector ones, we'll be stabilizing in another time.

I can't parse this comment, too many negations.^^ Integer vector registers can also cause ABI issues, if they are sometimes used by the calling convention but not always available.

RalfJung commented:
> Double negation in the first sentence is intended

Please avoid that, it makes things very hard to read and very easy to misunderstand.

"No ABI issues will not occur" means that ABI issues *will* occur which I hope is not what you meant.

> If variant calling convention is used in a function, standard ABI defines how to communicate between the function itself (callee) and the caller using vector registers but nothing more (and all preserved vector registers are callee-saved). I can't find a case where vector integer ops can cause ABI issues.

I don't know what the "variant calling convention" is. I also don't know the specifics of any of these extensions, or really much about RISC-V at all. I am just replying to a comment that seemed to somehow say that "float registers could cause a problem but integers cannot", which doesn't make sense to me. Any extensions that makes more registers available could become a problem if arguments start to be passed in those registers. If the registers are never used for arguments, then great, we have no poblem.

a4lg commented:
> Sorry for being so nitpicky here, but I'd rather not risk introducing some soundness issue over a misunderstanding caused by a language barrier. Better to double-check. :)

I completely agree that.  We'd better to discuss it.

RalfJung commented:
Thanks! I'm still confused about what's happening with integer vectors, but since they are not being stabilized here we don't have to resolve that now.

tmandry commented:
@rfcbot reviewed

a4lg committed:
Stabilize 29 RISC-V target features (`riscv_ratified_v2`)

This commit stabilizes RISC-V target features with following constraints:

*   Describes a ratified extension.
*   Implemented on Rust 1.88.0 or before.
    Waiting for four+ version cycles seems sufficient.
*   Does not disrupt current rustc's target feature + ABI handling.
    It excludes "E" and all floating point-arithmetic extensions.
    "Zfinx" family does not involve floating point registers but
    not stabilizing for now to avoid possible confusion between
    the "F" extension family.
*   Not vector-related (floating point and integer).
    While integer vector subsets should not cause any ABI issues
    (as they don't use ABI-dependent floating point registers),
    we need to discuss before stabilizing them.
*   Supported by the lowest LLVM version supported by rustc (LLVM 20).

List of target features to be stabilized:

1.  "b"
2.  "za64rs"   (no-RT)
3.  "za128rs"  (no-RT)
4.  "zaamo"
5.  "zabha"
6.  "zacas"
7.  "zalrsc"
8.  "zama16b"  (no-RT)
9.  "zawrs"
10. "zca"
11. "zcb"
12. "zcmop"
13. "zic64b"   (no-RT)
14. "zicbom"
15. "zicbop"   (no-RT)
16. "zicboz"
17. "ziccamoa" (no-RT)
18. "ziccif"   (no-RT)
19. "zicclsm"  (no-RT)
20. "ziccrse"  (no-RT)
21. "zicntr"
22. "zicond"
23. "zicsr"
24. "zifencei"
25. "zihintntl"
26. "zihintpause"
27. "zihpm"
28. "zimop"
29. "ztso"

Of which, 20 of them (29 minus 9 "no-RT" target features) support
runtime detection through `std::arch::is_riscv_feature_detected!()`.
